### PR TITLE
Pro 5369 solicitors as executor in grant flow

### PR DESCRIPTION
**Pull request by jackmunkonge**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRO-5369

### Change description ###
Solicitors as executor in grant flow

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/probate/controller/BusinessValidationController.java [hunk 1]
```diff
@@ -66,6 +66,11 @@
     private static final String DEFAULT_LOG_ERROR = "Case Id: {} ERROR: {}";
     private static final String INVALID_PAYLOAD = "Invalid payload";
 
+    @PostMapping(path = "/sols-apply-as-exec")
+    public ResponseEntity<CallbackResponse> solsApplyAsExec(@RequestBody CallbackRequest request) {
+        return ResponseEntity.ok(callbackResponseTransformer.toggleApplicantFields(request));
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/probate/controller/BusinessValidationController.java]: rename to setApplicantFieldsForSolsApplyAsExec

File: src/main/java/uk/gov/hmcts/probate/model/ccd/raw/request/CaseData.java [hunk 2]
```diff
@@ -552,9 +571,93 @@ private boolean isPrimaryApplicantNotApplying() {
...
+        }
+
         return totalExecutors.stream().filter(ex -> isApplying(ex, applying)).collect(Collectors.toList());
     }
 
+    private boolean isSolicitorCreatedGrant(String solsWillType) {
+        return (solsWillType != null);
+    }
+
+    private List<CollectionMember<AdditionalExecutor>> mapAdditionalExecutorsApplying(List<CollectionMember<AdditionalExecutorApplying>> additionalExecutors) {
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/probate/model/ccd/raw/request/CaseData.java:591]: please make sure additional functionality is covered by unit tests for this class

File: src/main/java/uk/gov/hmcts/probate/model/ccd/raw/request/CaseData.java [hunk 3]
```diff
@@ -552,9 +571,93 @@ private boolean isPrimaryApplicantNotApplying() {
...
+        List<CollectionMember<AdditionalExecutor>> newAdditionalExecutors = new ArrayList<>();
+
+        for (CollectionMember<AdditionalExecutorApplying> e : additionalExecutors) {
+            exec = e.getValue();
+            String forenames = exec.getApplyingExecutorFirstName();
+            String surname = exec.getApplyingExecutorLastName();
+
+            if (exec.getApplyingExecutorFirstName() == null || exec.getApplyingExecutorLastName() == null) {
+                List<String> names = splitFullname(exec.getApplyingExecutorName());
+                surname = names.remove(names.size()-1);
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/probate/model/ccd/raw/request/CaseData.java]: Check double barrelled sur names and hyphenated fore/surnames work as expected. Also single names (eg. Madonna)

File: src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java [hunk 4]
```diff
@@ -581,6 +593,10 @@ private boolean isCodicil(CaseData caseData) {
         return YES.equals(caseData.getWillHasCodicils());
     }
 
+    private boolean isApplyAsExecutor(CaseData caseData) {
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java]: isSolicitorApplyingAsExecutor

File: src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java [hunk 5]
```diff
@@ -831,6 +877,24 @@ private void updateCaseBuilderForTransformCase(CaseData caseData, ResponseCaseDa
                     .willNumberOfCodicils(null);
         }
 
+        if (isApplyAsExecutor(caseData)) {
+            if (YES.equals(caseData.getSolsSolicitorIsMainApplicant())) {
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java:877]: see if you can refactor this function into its own method for both calls

File: src/test/java/uk/gov/hmcts/probate/controller/BusinessValidationControllerTest.java [hunk 6]
```diff
@@ -190,6 +192,10 @@ public void setup() {
                 .solsSolicitorAddress(solsAddress)
                 .ihtFormId(IHT_FORM)
                 .solsSOTName(SOLICITOR_NAME)
+                .solsSolicitorIsApplyingExec(YES)
```
Comment (sanjayparekh-hmcts) [inline src/test/java/uk/gov/hmcts/probate/controller/BusinessValidationControllerTest.java:198]: check tests for these values being null or NO

**Review by sanjayparekh-hmcts — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java (hunk 5 above)
Comment (jackmunkonge) [reply src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java:877]: Ends up making code less readable; having to pass multiple mutable objects as arguments

**Review by sanjayparekh-hmcts — APPROVED** [review]

**Pull request merged**